### PR TITLE
Fix incompatible types in assignment

### PR DESCRIPTION
**Pull request by daikikatsuragawa**
Signed-off-by: Daiki Katsuragawa <50144563+daikikatsuragawa@users.noreply.github.com>

This pull request is one of the steps of #331 . Most of the "`Incompatible types in assignment`" errors detected by mypy are fixed. Basically, the policy is to modify the process without major changes. Therefore, there are also temporary responses (e.g., defining type hints `Any`). These are excluded from the modifications in this pull request and will be revised later.

In conclusion, this pull request will have the following effects.

- Deletion of "`Incompatible types in assignment`" processes
- Prevention of unexpected exceptions
- Improved readability with type hints

### CODE REVIEW

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 1]
```diff
@@ -343,12 +344,16 @@ def _predict_fn_custom(self, input_instance, desired_class):
 
     def compute_yloss(self, cfs, desired_range, desired_class):
         """Computes the first part (y-loss) of the loss function."""
-        yloss = 0.0
+        yloss: Any = 0.0
```
Comment (daikikatsuragawa) [inline dice_ml/explainer_interfaces/dice_genetic.py:347]: Currently, multiple types of values are stored.

File: dice_ml/explainer_interfaces/dice_pytorch.py [hunk 2]
```diff
@@ -223,14 +224,16 @@ def do_optimizer_initializations(self, optimizer, learning_rate):
...
+        self.optimizer: Optional[Union[torch.optim.Adam, torch.optim.RMSprop]] = None
         if opt_method == "adam":
             self.optimizer = torch.optim.Adam(self.cfs, lr=learning_rate)
         elif opt_method == "rmsprop":
             self.optimizer = torch.optim.RMSprop(self.cfs, lr=learning_rate)
 
     def compute_yloss(self):
         """Computes the first part (y-loss) of the loss function."""
-        yloss = 0.0
+        yloss: Any = 0.0
```
Comment (daikikatsuragawa) [inline dice_ml/explainer_interfaces/dice_pytorch.py:235]: Currently, multiple types of values are stored.

File: dice_ml/explainer_interfaces/dice_pytorch.py [hunk 3]
```diff
@@ -425,7 +428,7 @@ def find_counterfactuals(self, query_instance, desired_class, optimizer, learnin
         test_pred = self.predict_fn(torch.tensor(query_instance).float())[0]
         if desired_class == "opposite":
             desired_class = 1.0 - np.round(test_pred)
-        self.target_cf_class = torch.tensor(desired_class).float()
+        self.target_cf_class: Any = torch.tensor(desired_class).float()
```
Comment (daikikatsuragawa) [inline dice_ml/explainer_interfaces/dice_pytorch.py:431]: Currently, multiple types of values are stored.

File: dice_ml/explainer_interfaces/dice_tensorflow2.py [hunk 4]
```diff
@@ -341,8 +341,7 @@ def initialize_CFs(self, query_instance, init_near_query_instance=False):
                         one_init.append(np.random.uniform(self.minx[0][i], self.maxx[0][i]))
                 else:
                     one_init.append(query_instance[0][i])
-            one_init = np.array([one_init], dtype=np.float32)
-            self.cfs[n].assign(one_init)
+            self.cfs[n].assign(np.array([one_init], dtype=np.float32))
```
Comment (daikikatsuragawa) [inline dice_ml/explainer_interfaces/dice_tensorflow2.py:344]: Substitute directly. This prevents `one_init` from being of more than one type.

File: dice_ml/explainer_interfaces/feasible_model_approx.py [hunk 5]
```diff
@@ -81,8 +81,9 @@ def train(self, constraint_type, constraint_variables, constraint_direction, con
             train_loss = 0.0
             train_size = 0
 
-            train_dataset = torch.tensor(self.vae_train_feat).float()
-            train_dataset = torch.utils.data.DataLoader(train_dataset, batch_size=self.batch_size, shuffle=True)
+            train_dataset = torch.utils.data.DataLoader(
+                torch.tensor(self.vae_train_feat).float(), # type: ignore
```
Comment (daikikatsuragawa) [inline dice_ml/explainer_interfaces/feasible_model_approx.py:85]: Set `type: ignore` because the error depends on the implementation of `torch.utils.data.DataLoader`.

**Comment by amit-sharma** [comment]
@daikikatsuragawa thanks for this contribution. Can you share a few lines on the motivation for this PR (e.g., why change `cfs` from [] to `pd.DataFrame()`?) and are there any bugs that this PR fixes?

**Comment by daikikatsuragawa** [comment]
> @daikikatsuragawa thanks for this contribution. Can you share a few lines on the motivation for this PR (e.g., why change `cfs` from [] to `pd.DataFrame()`?) and are there any bugs that this PR fixes?

**Comment by daikikatsuragawa** [comment]
@amit-sharma 

Comment on each. After completion, send a message with a Mention.

**Comment by daikikatsuragawa** [comment]
Operational error for `closed` (and `reopened`). My apologies.

**Review by gaugup — CHANGES_REQUESTED** [review]
Should we enable mypy in dice-ml linting build and fix all the mypy errors?

**Comment by daikikatsuragawa** [comment]
@amit-sharma 
> Should we enable mypy in dice-ml linting build and fix all the mypy errors?

Yes. However, some mypy errors remain at present. It would be good to add them to Lint after they have been resolved.

**Comment by daikikatsuragawa** [comment]
@amit-sharma 

> @daikikatsuragawa thanks for this contribution. Can you share a few lines on the motivation for this PR (e.g., why change cfs from [] to pd.DataFrame()?)

During the function, the process immediately following was checked and it was determined that the type of `cfs` is `pd.DataFrame`. If this is not appropriate, please let me know.

https://github.com/interpretml/DiCE/blob/490b7abce937b2008cf4b7b5dd6ca8f579f967f7/dice_ml/explainer_interfaces/dice_KD.py#L173-L176

> and are there any bugs that this PR fixes?

Probably this case (`cfs = pd.DataFrame()`) only. The others are fixes related to type hinting and refactoring and are not expected to affect behaviour. Unit tests also passed.
